llvm/lib/IR/Instruction.cpp

```diff
@@ -212,14 +208,13 @@ void Instruction::moveBeforeImpl(BasicBlock &BB, InstListType::iterator I,
   BB.getInstList().splice(I, getParent()->getInstList(), getIterator());
 
   if (BB.IsNewDbgInfoFormat && !Preserve) {
-    if (!DbgMarker)
-      BB.createMarker(this);
     DPMarker *NextMarker = getParent()->getNextMarker(this);
 
     // If we're inserting at point I, and not in front of the DPValues attached
     // there, then we should absorb the DPValues attached to I.
-    if (NextMarker && !InsertAtHead)
-      DbgMarker->absorbDebugValues(*NextMarker, false);
+    if (!InsertAtHead && NextMarker && !NextMarker->empty()) {
+      adoptDbgValues(&BB, I, false);
```

Comment:
Given that (most? all?) other uses of `adoptDbgValues` involve a check-if-not-then-cleanup-something, can this get a comment explaining why it isn't needed?

Do we want to make `adoptDbgValues` nodiscard, or is that too severe?

+1 to what @SLTozer said:

> it's not easy to intuit what's happening in each case and what the difference is without really digging into it ... [can  we] extract common behaviour out to other functions [in a future patch].

Reply:
Hmmm, I was going to say "because we never have trailing DPValues here" (i.e. ones hanging off the end of the block), but I'm now seeing various scenarios where moveBefore is called with the end() iterator. And the assertion on moveBeforeImpl's entry explicitly takes account of that. So this would need some cleanup.

It's probably better to fold all that maintenance code into adoptDbgValues, I suppose I'll look at that next.